### PR TITLE
MeshPhysicalMaterial: Added sheen

### PR DESCRIPTION
**Pull request by DanielSturk**
Added sheen for cloth materials such as velvet. The BDRF used is the same as Google's Filament (NDF is Estevez's and Kulla's "Charlie" sheen + Neubelt's visibility term https://google.github.io/filament/Filament.md.html).

I also added support for sheen to the StandardNodeMaterial (@sunag)

The demo for testing is webgl_materials_sheen.html

### CODE REVIEW

**Comment by DanielSturk** [comment]
By the way, this implementation only covers direct lighting for the time being. There's a bit of confusion/complication in regards to IBL and ambient lighting because sheen seems to be dependent on the direction of the light source

**Comment by bhouston** [comment]
We should never support ambient lighting on this.

**Comment by sunag** [comment]
Thank you for your contribution once again...
The parameter and example `sheen` does not seem to be working.
I believe that changes in the `core` and in the `NodeMaterial` should be sent separately but I think if possible unify this feature using nodes only without modify the core...

**Comment by DanielSturk** [comment]
@sunag

> The parameter and example `sheen` does not seem to be working.

What do you mean by this? I asked my colleague tested it and it runs fine on his machine

**Comment by sciecode** [comment]
@sunag 

If you are using something like "githack" to test, it won't work. 
`Three.module.js` needs to be updated, because of new `sheen` parameter. Running it locally works as intended.

**Comment by bhouston** [comment]
> unify this feature using nodes only without modify the core...

The reason we are adding this is that it is a BRDF improvement and it is to align with the Enterprise PBR model:  https://github.com/DassaultSystemes-Technology/EnterprisePBRShadingModel

**Comment by bhouston** [comment]
This is being done in this later context of adopting PBR Next / Enterprise PBR https://github.com/mrdoob/three.js/issues/16977 and ensuring Node-graph materials are awesome https://github.com/mrdoob/three.js/issues/16440

**Comment by sunag** [comment]
@DanielSturk @sciecode Yeah, I run locally and work thanks.

> The reason we are adding this is that it is a BRDF improvement and it is to align with the Enterprise PBR model: https://github.com/DassaultSystemes-Technology/EnterprisePBRShadingModel

The roadmap is very good. My think is advance in BRDF using nodes only to create better support for it and facilitate the merger of PR. Once implemented, create other BRDF features will be easier and optimized since the code can be calculated modularly. `Sheen` could be one of these new BRDF nodes.

**Comment by bhouston** [comment]
Usually the brdf parameters are part of the base brdf node.  Having brdf
parameters as separate nodes generally is not how brdfs are implemented.

**Comment by sunag** [comment]
> Usually the brdf parameters are part of the base brdf node.  Having brdf
parameters as separate nodes generally is not how brdfs are implemented.

I was not thinking about separating the code from BRDF class, just implementing it as a process tree a bit more organized.
`brdf/SheenNode` for example...

So we can add relatives `sheen` functions if the parameter `.sheen` is really used.
Relative the performance, it is good too, even unused function adds an increased time to `program` creation.

**Comment by sunag** [comment]
@mrdoob @WestLangley I do not want to block this PR, maybe I have caused this impression, it looks good to me. I can improve these details relative to the `NodeMaterial` later with other PR.

**Comment by DanielSturk** [comment]
> @mrdoob @WestLangley I do not want to block this PR, maybe I have caused this impression, it looks good to me. I can improve these details relative to the `NodeMaterial` later with other PR.

File: examples/webgl_materials_sheen.html
Comment (Mugen87) [inline examples/webgl_materials_sheen.html]: Can you please add this example to `files.js`?  Otherwise it won't be visible in the side bar.

**Comment by Mugen87** [comment]
https://raw.githack.com/DanielSturk/three.js/sheen/examples/webgl_materials_sheen.html

It seems it's not possible to control the arm with the keys like in the original [physics example](https://raw.githack.com/DanielSturk/three.js/sheen/examples/webgl_physics_cloth.html). Besides, the arm jitters right now in a strange way. At least on my system.

**Comment by yomboprime** [comment]
> https://raw.githack.com/DanielSturk/three.js/sheen/examples/webgl_materials_sheen.html
> 
> It seems it's not possible to control the arm with the keys like in the original [physics example](https://raw.githack.com/DanielSturk/three.js/sheen/examples/webgl_physics_cloth.html). Besides, the arm jitters right now in a strange way. At least on my system.
> 
> If no interaction is intended, consider to remove the arm or make it static.

He has put a `Math.random()` in the arm control.

File: examples/webgl_materials_sheen.html
Comment (DanielSturk) [reply examples/webgl_materials_sheen.html]: Done

**Comment by DanielSturk** [comment]
@Mugen87 @yomboprime I didn't expect the arm to be so controversial :p
My coworker made us this model to use instead
126e79a1512777e4966ebfeef3b0ebe7

**Comment by Mugen87** [comment]
That's better^^! 

BTW: It seems that changing the `sheen` via `dat.gui` does not have a visual effect. Or do I miss something?

File: src/materials/MeshPhysicalMaterial.js
Comment (Mugen87) [inline src/materials/MeshPhysicalMaterial.js]: It will be necessary to enhance the docs and TypeScript declaration files with this parameter. However, you can do this with a different PR after this one gets merged.

**Comment by EliasHasle** [comment]
> BTW: It seems that changing `sheen` via `dat.gui` does not have a visual effect. Or do I miss something?

@DanielSturk That's right. The gui modifies properties of a separate params object and has no event handlers. The simplest solution will be something like:

    gui.add(material.uniforms.someFloatUniform, "value", 0, 1);

**Comment by DanielSturk** [comment]
@EliasHasle
I'm almost certain the githack demo is broken because you're not supposed to submit the build files in PRs.

**Comment by EliasHasle** [comment]
OK. Then I recommend that you do like me and create a personal github pages repository for these kinds of demos, where you can control which files are accessed. I would very much like to see this new velvet beast, or whatever it is. 😄

**Comment by DanielSturk** [comment]
> OK. Then I recommend that you do like me and create a personal github pages repository for these kinds of demos

@EliasHasle  Great idea... instead I just created another personal branch for this demo :p

https://raw.githack.com/DanielSturk/three.js/sheen-demo/examples/webgl_materials_sheen.html

**Comment by EliasHasle** [comment]
That is an impressive demo! Wow! :D

**Comment by DanielSturk** [comment]
> That is an impressive demo! Wow!

Thank you!

**Review by Mugen87 — APPROVED** [review]

File: src/materials/MeshPhysicalMaterial.js
Comment (DanielSturk) [reply src/materials/MeshPhysicalMaterial.js]: Done

**Comment by Mugen87** [comment]
@DanielSturk  Can you please resolve the merge conflicts?

**Comment by Mugen87** [comment]
@mrdoob Looks ready to merge 👍

**Comment by bhouston** [comment]
Is there a reason this isn't being merged?

**Comment by Mugen87** [comment]
Sorry, I'm sure this PR was just overlooked. 

@DanielSturk Can you please resolve the merge conflicts one more time?

**Comment by romainguy** [comment]
@DanielSturk What problems did you run into with IBL support for `sheen`? Folks on Babylon.js and I both got it to work, we can help if you'd like.

**Comment by DanielSturk** [comment]
@Mugen87 Np. Merge conflicts resolved :)

**Comment by DanielSturk** [comment]
> @DanielSturk What problems did you run into with IBL support for `sheen`? Folks on Babylon.js and I both got it to work, we can help if you'd like.

Thank you, I appreciate the support! I've got it working now (was a little confused by the DFG (DG) LUT but it's working now). As soon as we can get this merged, I'll create the PR for the IBL Sheen!

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
@DanielSturk @bhouston I have renamed the property to `sheen` to match `emissive`, `specular`, ... https://github.com/mrdoob/three.js/commit/3e9af1f44514f3bdf2d25d21757bd637498b83a1

**Comment by mrdoob** [comment]
@sunag on one of my computers (MacBook Air, Early 2015, Intel HD Graphics 6000) the [sheen example](https://raw.githack.com/mrdoob/three.js/dev/examples/#webgl_materials_sheen) runs at 20fps when using the node material, and at 60fps when not using it.

In this screenshot, the fps graph shows when nodematerial was enabled and disabled:

Screenshot 2019-08-20 at 17 10 59

**Comment by sunag** [comment]
> @sunag on one of my computers (MacBook Air, Early 2015, Intel HD Graphics 6000) the sheen example runs at 20fps when using the node material, and at 60fps when not using it.

This because the NodeMaterial is being compiled every frame: `nodeMaterial.needsCompile = true;` is added in render loop. I already send a PR to fix this.

**Comment by mrdoob** [comment]
> This because the NodeMaterial is being compiled every frame: `nodeMaterial.needsCompile = true;` is added in render loop. I already send a PR to fix this.

Oh, what's the PR that fixes it?

**Comment by sunag** [comment]
> Oh, what's the PR that fixes it?

Ops, I'm already sending, sry my english :flushed:

**Comment by DanielSturk** [comment]
Thanks for the merge @mrdoob and everyone involved
Make sure to check out my follow up PR for IBL support @romainguy https://github.com/mrdoob/three.js/pull/17303

**Comment by donmccurdy** [comment]
The glTF sheen extension ([KHR_materials_sheen](https://github.com/KhronosGroup/glTF/pull/1688)) is nearly complete, although still welcoming feedback for a bit longer. To implement in MeshPhysicalMaterial, it would require:

- sheen
- sheenMap
- sheenRoughness
- sheenRoughnessMap

The two maps use `.rgb` and `.a` channels respectively; they can be combined.

**Pull request merged**
